### PR TITLE
randomized network params & batch runner

### PR DESCRIPTION
**Pull request by yusefnapora**
This adds parameters for the testground `network.LinkShape` fields that let you specify a range of values as a two-element JSON array of either floats or duration strings (depending on the parameter type). In service of https://github.com/filecoin-project/oni/issues/14 

For example:

```toml

latency_range = '["20ms", "500ms"]'
loss_range = '[0, 0.2]'
```

Would pick a random latency between 20ms and 500ms, and a packet loss probability between 0 and 0.2

The `ApplyNetworkParameters` function will choose a random value within the range and set the configuration, and it dumps the `LinkShape` as a JSON object to `network-link-shape.json` in the testground assets dir for future reference / analysis.

I also added a `runner` command that will run a composition multiple times and collect the outputs. E.g.:

```
go run ./runner -runs 5 _compositions/net-chaos/latency.toml
```

There's a `-parallel <n>` flag that will try to run multiple jobs in parallel, but it doesn't work with the `local:docker` runner (fails with `Pool overlaps with other one on this address space` error).  I haven't got a k8s cluster to test with yet, but maybe @nonsense can try it. It's probably only useful on k8s anyway, so I don't really care about local:docker.

### CODE REVIEW

**Comment by nonsense** [comment]
LGTM, will try it out and report back.

File: lotus-soup/testkit/testenv.go [hunk 1]
```diff
@@ -32,10 +34,39 @@ func (t *TestEnvironment) DurationParam(name string) time.Duration {
...
 	t.RecordMessage(spew.Sprintf(format, args...))
 }
 
+func (t *TestEnvironment) DumpJSON(filename string, v interface{}) {
+	b, err := json.Marshal(v)
+	if err != nil {
+		t.RecordMessage("unable to marshal object to JSON: %s", err)
+		return
+	}
+	f, err := t.CreateRawAsset(filename)
```
Comment (raulk) [inline lotus-soup/testkit/testenv.go:59]: missing `defer f.Close()`

File: lotus-soup/testkit/testenv.go [hunk 2]
```diff
@@ -49,6 +80,72 @@ func WrapTestEnvironment(f func(t *TestEnvironment) error) run.InitializedTestCa
...
 		t := &TestEnvironment{RunEnv: runenv, InitContext: initCtx}
 		t.Role = t.StringParam("role")
+
+		t.DumpJSON("test-parameters.json", t.TestInstanceParams)
+
 		return f(t)
 	}
 }
+
+// parameter types (TODO: move somewhere sensible)
```
Comment (raulk) [inline lotus-soup/testkit/testenv.go:90]: nit: Maybe move the ranges types to a `testenv_ranges.go` file?

File: lotus-soup/runner/main.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	log.Printf("starting test run %d. writing testground client output to %s\n", job.runNumber, outPath)
+	if err = cmd.Run(); err != nil {
+		return jobResult{job: job, runError: err}
+	}
+	return jobResult{job: job}
+}
+
+func worker(id int, jobs <-chan jobDefinition, results chan<- jobResult) {
+	for j := range jobs {
```
Comment (raulk) [inline lotus-soup/runner/main.go:52]:
```suggestion
    log.Printf("started worker %d", id)
	for j := range jobs {
```

Comment (raulk) [reply lotus-soup/runner/main.go:52]: Let's log that we started a worker.

File: lotus-soup/runner/main.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		log.Printf("worker %d started test run %d", id, j.runNumber)
```
Comment (nonsense) [inline lotus-soup/runner/main.go:53]:
```suggestion
		log.Printf("worker %d started test run %d\n", id, j.runNumber)
```

File: lotus-soup/runner/main.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		dir := path.Join(outdir, fmt.Sprintf("run-%d", j))
+		skipStdout := *parallelism != 1
+		jobs <- jobDefinition{runNumber: j, compositionPath: compositionPath, outputDir: dir, skipStdout: skipStdout}
+	}
+	close(jobs)
+
+	for i := 0; i < *runs; i++ {
+		r := <-results
+		if r.runError != nil {
+			log.Printf("error running job %d: %s", r.job.runNumber, r.runError)
```
Comment (nonsense) [inline lotus-soup/runner/main.go:116]:
```suggestion
			log.Printf("error running job %d: %s\n", r.job.runNumber, r.runError)
```

**Comment by raulk** [comment]
@yusefnapora ping! I think we're good to merge, but there are a few comments/nits above.

File: lotus-soup/testkit/testenv.go (hunk 2 above)
Comment (raulk) [reply lotus-soup/testkit/testenv.go]: Took care of it.

**Comment by raulk** [comment]
Enables #14 and #26.

**Pull request merged**